src/connectivity/percolation.c

```diff
@@ -0,0 +1,290 @@
+/*
+    IGraph library.
+    Copyright (C) 2025  The igraph development team <igraph@igraph.org>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_components.h"
+#include "igraph_constants.h"
+#include "igraph_error.h"
+#include "igraph_interface.h"
+#include "igraph_types.h"
+#include "igraph_vector.h"
+
+/**
+ * \function igraph_i_percolate_edge
+ * \brief Percolates a single edge.
+ * \param links Vector representing parents.
+ * \param sizes sizes[i] is the number of children of links[i]
+ * \param biggest The biggest value in sizes, is updated if a bigger cluster is created.
+ * \param a A vertex incident to the edge.
+ * \param b The other vertex incident to the edge.
+ */
+
+static void percolate_edge(igraph_vector_int_t *links,
+                           igraph_vector_int_t *sizes,
+                           igraph_integer_t *biggest,
+                           igraph_integer_t a,
+                           igraph_integer_t b) {
+
+    // Find head of each tree
+    while (VECTOR(*links)[a] != a) {
+        VECTOR(*links)[a] = VECTOR(*links)[VECTOR(*links)[a]];
+        a = VECTOR(*links)[a];
+    }
+    while (VECTOR(*links)[b] != b) {
+        VECTOR(*links)[b] = VECTOR(*links)[VECTOR(*links)[b]];
+        b = VECTOR(*links)[b];
+    }
+
+    // If they are already connected, exit early
+    if (a == b) {
+        return;
+    }
+
+    // Make smaller child of larger
+    igraph_integer_t parent, child;
+    if (VECTOR(*sizes)[a] < VECTOR(*sizes)[b]) {
+        parent = b;
+        child = a;
+    } else {
+        parent = a;
+        child = b;
+    }
+
+    // Make a child of b
+    VECTOR(*links)[child] = parent;
+    VECTOR(*sizes)[parent] += VECTOR(*sizes)[child];
+
+    // If made new biggest component, update biggest
+    if (VECTOR(*sizes)[parent] >= *biggest) {
+        *biggest = VECTOR(*sizes)[parent];
+    }
+}
+
+/**
+ * \function igraph_i_edge_list_percolation
+ * \brief Gives the size of the largest connected component as edges are added.
+ * \param edges Vector of edges, will be added in order.
+ * \param output output[i] is the size of the largest connected component after edges[i] is added
+ */
+
+static igraph_error_t edge_list_percolation(
```

Comment:
Let's make this function public (including adding it to headers, adding docs, renaming, adding to interfaces).